### ISSUE TITLE
Swift Port Architecture: Multi-Camera Intelligent Vision System for iPhone + Smart Glasses

### ISSUE THREAD

**Issue by ebowwa** [dako2/edaxshifu#12]
# Swift Port Architecture for LiveLearningCamera

## Overview
Port of the Python-based EdaxShifu intelligent vision system to native iOS/Swift, with support for multiple camera sources including iPhone cameras and smart glasses.

### Core Components

1. **YOLO Detector** (`yolo_detector.py`)
   - ONNX-based YOLOv11 object detection
   - 80 COCO classes recognition
   - Hand-object interaction detection
   - Non-max suppression for duplicate removal

2. **KNN Classifier** (`knn_classifier.py`)
   - ResNet18 feature extraction (512-dim embeddings)
   - Few-shot learning with cosine similarity
   - Adaptive learning from feedback
   - Memory management (100 samples/class max)
   - Model persistence

3. **Intelligent Capture** (`intelligent_capture.py`)
   - RTSP/webcam stream integration
   - Trigger system (keyboard, object detection)
   - Live model reloading
   - Success/failure feedback loop
   - Gemini API fallback for unknown objects

4. **Unified Interface** (`unified_interface.py`)
   - Gradio web UI
   - Live streaming view
   - Capture & annotation
   - Teaching interface
   - Real-time stats

### Python Dependencies
- OpenCV (camera/video processing)
- PyTorch & torchvision (ResNet18 feature extraction)
- scikit-learn (KNN classifier)
- NumPy (array operations)
- Gradio (web UI)
- ONNX Runtime (YOLO inference)

### Phase 1: Core Infrastructure (Week 1-2)
- [ ] AVFoundation camera setup
- [ ] Vision framework integration
- [ ] Core ML model management
- [ ] SwiftUI camera preview

### Phase 2: Object Detection (Week 2-3)
- [ ] Convert YOLO ONNX to Core ML
- [ ] VNCoreMLRequest implementation
- [ ] Bounding box processing
- [ ] NMS implementation
- [ ] Real-time overlay rendering

### Phase 3: KNN Classification (Week 3-4)
- [ ] ResNet18 Core ML conversion
- [ ] Feature extraction pipeline
- [ ] Swift KNN implementation
- [ ] Embedding storage (Core Data)
- [ ] Cosine similarity metrics

### Phase 4: Intelligent Capture (Week 4-5)
- [ ] Gesture recognizers
- [ ] Object detection triggers
- [ ] Frame capture pipeline
- [ ] Success/failure tracking
- [ ] Feedback loop implementation

### Phase 5: Data Persistence (Week 5-6)
- [ ] Core Data schema
- [ ] CloudKit sync (optional)
- [ ] Model versioning
- [ ] Training sample management
- [ ] Export/import capabilities

### Phase 6: UI/UX (Week 6-7)
- [ ] Camera view with overlays
- [ ] Capture & annotation UI
- [ ] Teaching interface
- [ ] Statistics dashboard
- [ ] Settings & configuration

### Phase 7: Advanced Features (Week 7-8)
- [ ] Live learning updates
- [ ] Background processing
- [ ] Share extensions
- [ ] Widget support
- [ ] Shortcuts integration

## Python to Swift/iOS Component Mapping

| Python Component | Swift/iOS Equivalent | Notes |
|-----------------|---------------------|-------|
| OpenCV | AVFoundation + Vision | Native camera & image processing |
| PyTorch/ResNet18 | Core ML + CreateML | Pre-trained models & feature extraction |
| ONNX Runtime | Core ML | Model inference engine |
| scikit-learn KNN | Accelerate/Metal | Custom implementation with SIMD |
| NumPy arrays | Swift Arrays + Accelerate | Vectorized operations |
| Gradio UI | SwiftUI | Native iOS interface |
| RTSP streaming | AVPlayer/Network.framework | Network video streams |
| Threading | GCD/Combine | Concurrency & async operations |
| Pickle/NPZ | Core Data/JSON | Model persistence |
| PIL/Image | UIImage/CGImage | Image manipulation |

### Vision
The iPhone acts as the central AI brain while supporting multiple camera sources - built-in cameras, smart glasses, action cameras, or any network-connected vision device.

### Camera Source Abstraction

```swift
protocol CameraSource {
    var sourceID: String { get }
    var sourceName: String { get }
    var sourceType: CameraSourceType { get }
    var isConnected: Bool { get }
    var capabilities: CameraCapabilities { get }
    
    func startStream() async throws -> AsyncStream<CameraFrame>
    func captureHighResPhoto() async throws -> UIImage
    func getMetadata() -> CameraMetadata
}

enum CameraSourceType {
    case builtIn(position: AVCaptureDevice.Position)
    case smartGlasses(model: GlassesModel)
    case networkCamera(protocol: StreamProtocol)
    case externalUSBC
    case wirelessCamera(connection: ConnectionType)
}
```

### Supported Smart Glasses
1. **Ray-Ban Meta** - WiFi Direct streaming
2. **XREAL Air** - USB-C/Lightning connection
3. **Apple Vision Pro** - Continuity Camera API
4. **Custom BLE Glasses** - ESP32-CAM based

### Communication Protocols
- BLE for control & metadata
- Network.framework for video streaming
- MultipeerConnectivity for P2P connections
- WiFi Direct for high-bandwidth streaming

#### 1. Stereoscopic Depth & 3D Understanding
- Calculate 3D positions from multiple viewpoints
- Depth-aware object detection
- Spatial relationship understanding

#### 2. 360° Awareness
- iPhone rear camera: forward view
- Smart glasses: POV/peripheral view
- iPhone front camera: self/gesture view

#### 3. Context-Aware Processing
```swift
enum CaptureContext {
    case handheld      // Using iPhone camera
    case wearable      // Using glasses as primary
    case surveillance  // Multiple fixed cameras
    case collaborative // Multiple users/devices
}
```

#### 1. Gesture Recognition from POV
- Detect hand gestures in first-person view
- Trigger captures with hand signals
- Point-to-identify objects

#### 2. Gaze Tracking Integration
- Capture objects of interest based on gaze
- Prioritize detections in gaze area
- Attention-based learning

#### 3. Contextual AR Overlays
- Minimal UI rendered on glasses display
- Object labels and confidence scores
- Navigation and guidance arrows

### Distributed Processing Architecture

```swift
enum ProcessingMode {
    case onDevice      // All on iPhone
    case edgeCompute   // Glasses pre-process
    case cloudAssisted // Offload to cloud
    case collaborative // Multiple devices cooperate
}
```

### Use Cases

1. **Shopping Assistant**
   - Product identification and price comparison
   - Dietary restriction alerts
   - Shopping list tracking

2. **Learning & Education**
   - Equipment identification in labs
   - Procedure recording and playback
   - AR instruction overlays

3. **Professional Documentation**
   - Hands-free repair documentation
   - Automatic report generation
   - Step-by-step procedure capture

4. **Accessibility Enhancement**
   - Obstacle detection and navigation
   - Real-time text reading
   - Face recognition with names

5. **Social & Creative**
   - Multi-angle event capture
   - POV + third-person storytelling
   - Automatic best-shot selection

### Model Conversion
```bash
# YOLO: ONNX to Core ML
python -m coremltools.converters.onnx --model yolo11n.onnx

# ResNet18: PyTorch to Core ML
python convert_resnet.py
```

### Performance Optimizations
- Metal Performance Shaders for KNN
- Frame skipping for real-time processing
- Background queues for feature extraction
- Memory-mapped Core Data for large datasets

### iOS-Specific Enhancements
- Live Text integration for OCR
- Siri Shortcuts for quick capture
- iCloud sync for models/data
- App Clips for lightweight capture
- ARKit integration for spatial awareness

### Architecture Patterns
- MVVM with Combine
- Repository pattern for data
- Dependency injection
- Protocol-oriented design

### Testing Strategy
- XCTest for unit tests
- UI testing with XCUITest
- Performance testing with Instruments
- TestFlight beta testing

## Privacy & Security Considerations
- Per-camera permission management
- Encrypted streaming from glasses
- Face blurring options
- Recording indicators
- Data retention policies

## Next Steps
1. Set up Xcode project structure
2. Convert YOLO and ResNet18 models to Core ML
3. Implement camera source abstraction
4. Build basic capture pipeline
5. Add smart glasses connectivity
6. Develop multi-camera UI
7. Implement KNN classifier
8. Add teaching and feedback features

## Technical Requirements
- iOS 16.0+ (for latest Vision APIs)
- iPhone 12+ (for Neural Engine)
- Xcode 15+
- Swift 5.9+
- Core ML 5+

## References
- [Core ML Documentation](https://developer.apple.com/documentation/coreml)
- [Vision Framework](https://developer.apple.com/documentation/vision)
- [AVFoundation Camera](https://developer.apple.com/documentation/avfoundation/capture_setup)
- [MultipeerConnectivity](https://developer.apple.com/documentation/multipeerconnectivity)
